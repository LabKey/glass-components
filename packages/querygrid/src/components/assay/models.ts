/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, OrderedMap, Map, Record } from 'immutable'
import { AppURL, AssayDefinitionModel, AssayUploadTabs, QueryColumn, QueryInfo, FileAttachmentFormModel } from "@glass/base";

export interface AssayPropertiesPanelProps {
    model: AssayWizardModel
    onChange: Function
}

export interface IAssayURLContext {
    assayRequest?: string
    location?: string
    protocol: string
    provider: string
    runId?: string
}

export class AssayWizardModel extends Record({
    assayDef: undefined,
    isError: undefined,
    isWarning: false,
    isInit: false,
    isLoading: false,
    batchId: undefined, // batchId is null for first run
    lastRunId: undefined,
    returnURL: undefined,
    isSubmitted: undefined,
    isSubmitting: undefined,
    errorMsg: undefined,

    attachedFiles: Map<string, File>(),

    batchColumns: OrderedMap<string, QueryColumn>(),
    batchProperties: Map<string, any>(),
    runColumns: OrderedMap<string, QueryColumn>(),
    runId: undefined,
    runProperties: Map<string, any>(),
    runName: undefined,
    comment: undefined,
    dataText: undefined,
    queryInfo: QueryInfo,
    toDelete: undefined,
    selectedSamples: undefined,
}) implements FileAttachmentFormModel {
    assayDef: AssayDefinitionModel;
    isError?: boolean;
    isInit: boolean;
    isLoading: boolean;
    isWarning?: boolean;
    batchId? : number;
    lastRunId?: number;
    returnURL?: AppURL;
    isSubmitted?: boolean;
    isSubmitting?: boolean;
    errorMsg?: string;

    attachedFiles: Map<string, File>;

    batchColumns: OrderedMap<string, QueryColumn>;
    batchProperties: Map<string, any>;
    runColumns: OrderedMap<string, QueryColumn>;
    runId?: string;
    runProperties?: Map<string, any>;
    runName?: string;
    comment?: string;

    dataText: string;
    queryInfo: QueryInfo;
    toDelete?: string;
    selectedSamples?: Map<string, any>;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    // TODO shouldn't this move to actions.ts? or put just in Biologics somewhere
    static parseURLContext(location: any, params: any): IAssayURLContext {
        const { protocol, provider } = params;

        let context: IAssayURLContext = {
            location,
            protocol,
            provider
        };

        if (location && location.query) {
            const { assayRequest, runId } = location.query;
            context.assayRequest = assayRequest;
            context.runId = runId;
        }

        return context;
    }

    getAttachedFiles(): List<File> {
        return this.attachedFiles.valueSeq().toList();
    }

    getRunName(currentStep: AssayUploadTabs): string {
        if (this.runName) {
            return this.runName;
        }

        if (currentStep === AssayUploadTabs.Files) {
            // using file upload tab
            const file = this.getAttachedFiles().first();
            if (file) {
                return file.name;
            }
        }

        const date = new Date();
        const dateStr = date.toISOString().split('T')[0];
        let timeStr = date.toTimeString().split(' ')[0];
        timeStr = timeStr.replace(/:/g, '-');
        return this.assayDef.name + '_' + dateStr + '_' + timeStr;
    }
}