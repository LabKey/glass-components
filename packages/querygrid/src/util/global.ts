import createHistory from 'history/createBrowserHistory'

import { getGlobal, setGlobal } from "reactn";

/**
 * Initialize the global state object for this package.
 */
export function initBrowserHistoryState() {
    if (!getGlobal().utils) {
        setGlobal( {
            utils: {
                browserHistory: createHistory()
            }
        });
    }
}

/**
 * Access method for better browser history object from global state
 */
export function getBrowserHistory() {
    if (!getGlobal().utils || !getGlobal().utils.browserHistory) {
        throw new Error('Must call initBrowserHistoryState before you can access the global.utils.browserHistory object.');
    }

    return getGlobal().utils.browserHistory;
}