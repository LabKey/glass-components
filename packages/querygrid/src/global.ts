/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { getGlobal, setGlobal } from 'reactn'
import { List, Map } from 'immutable'
import { GRID_CHECKBOX_OPTIONS, QueryColumn, QueryGridModel, SchemaQuery, resolveSchemaQuery } from '@glass/base'

import { initBrowserHistoryState } from './util/global'
import { DataViewInfo, EditorModel, LookupStore } from './models'

/**
 * Initialize the global state object for this package.
 * @param metadata Optional Map to set the query metadata for this application
 * @param columnRenderers Optional Map to set the column renderers for this application
 */
export function initQueryGridState(metadata?: Map<string, any>, columnRenderers?: Map<string, any>) {
    if (!getGlobal().QueryGrid_models) {
        resetQueryGridState()
    }

    initBrowserHistoryState();

    if (metadata) {
        setQueryMetadata(metadata);
    }
    if (columnRenderers) {
        setQueryColumnRenderers(columnRenderers);
    }
}

/**
 * Clear out all of the global state object for this package
 */
export function resetQueryGridState() {
    setGlobal({
        QueryGrid_charts: Map<string, List<DataViewInfo>>(),
        QueryGrid_editors: Map<string, EditorModel>(),
        QueryGrid_lookups: Map<string, LookupStore>(),
        QueryGrid_metadata: Map<string, any>(),
        QueryGrid_models: Map<string, QueryGridModel>(),
        QueryGrid_columnrenderers: Map<string, any>()
    });
}

function getGlobalState(property: string) {
    if (!getGlobal()['QueryGrid_' + property]) {
        throw new Error('Must call initQueryGridState before you can access anything from the global.QueryGrid_' + property + ' objects.');
    }

    return getGlobal()['QueryGrid_' + property];
}

/**
 * Get the latest QueryGridModel object from the global state for a given modelId.
 * @param modelId QueryGridModel id to fetch
 */
export function getQueryGridModel(modelId: string): QueryGridModel {
    return getGlobalState('models').get(modelId);
}

export function getQueryGridModelsForSchema(schemaName: string): List<QueryGridModel> {
    return getGlobalState('models').filter(model => model.schema.toLowerCase() === schemaName.toLowerCase()).toList();
}

export function getQueryGridModelsForSchemaQuery(schemaQuery: SchemaQuery): List<QueryGridModel> {
    const modelName = resolveSchemaQuery(schemaQuery);
    return getGlobalState('models').filter(model => model.getModelName() === modelName).toList();
}

export function getQueryGridModelsForGridId(gridId: string): List<QueryGridModel> {
    const prefix = (gridId + '|').toLowerCase();
    return getGlobalState('models').filter(model => model.getId().indexOf(prefix) === 0).toList();
}

/**
 * Helper function for all callers/actions that would like to update information for a QueryGridModel in the global state.
 * @param model QueryGridModel in the global state to be updated, or to be added to global state if it does not already exist by Id
 * @param updates JS Object with the key/value pairs for updates to make to the model
 * @param connectedComponent Optional React.Component which should be re-rendered with this QueryGridModel update (prevents the need to "connect" the component to the global state)
 * @param failIfNotFound Boolean indicating if an error should be thrown if the model is not found in global state
 */
export function updateQueryGridModel(model: QueryGridModel, updates: any, connectedComponent?: React.Component, failIfNotFound: boolean = true): QueryGridModel {
    if (failIfNotFound && !getGlobalState('models').has(model.getId())) {
        throw new Error('Unable to find QueryGridModel for modelId: ' + model.getId());
    }

    const updatedModel = model.merge(updates) as QueryGridModel;

    setGlobal({
        QueryGrid_models: getGlobalState('models').set(model.getId(), updatedModel)
    },

    (global) => {
        if (connectedComponent) {
            connectedComponent.forceUpdate();
        }
    });

    return updatedModel;
}

/**
 * Remove a QueryGridModel from the global state
 * @param model QueryGridModel to be removed
 * @param connectedComponent Optional React.Component which should be re-rendered with this QueryGridModel update (prevents the need to "connect" the component to the global state)
 */
export function removeQueryGridModel(model: QueryGridModel, connectedComponent?: React.Component) {
    setGlobal({
        QueryGrid_models: getGlobalState('models').delete(model.getId()),
        QueryGrid_editors: getGlobalState('editors').delete(model.getId())
    },

    (global) => {
        if (connectedComponent) {
            connectedComponent.forceUpdate();
        }
    });
}

/**
 * Get the query metadata object from the global QueryGrid state
 */
export function getQueryMetadata() {
    return getGlobalState('metadata');
}

/**
 * Sets the query metadata object to be used for this application in the global QueryGrid state
 * @param metadata Map of query metadata to be applied to the query infos and column infos
 */
export function setQueryMetadata(metadata: Map<string, any>) {
    setGlobal({
        QueryGrid_metadata: metadata
    });
}

/**
 * Get the query grid column renderers map from the global QueryGrid state
 */
export function getQueryColumnRenderers() {
    return getGlobalState('columnrenderers');
}

/**
 * Sets the valid column renderers for this application in the global QueryGrid state
 * @param renderers Map of query grid column renderers to be bound to the queryInfo columns
 */
export function setQueryColumnRenderers(columnrenderers: Map<string, any>) {
    setGlobal({
        QueryGrid_columnrenderers: columnrenderers
    });
}

/**
 * Get the list of DataViewInfos from the global state for a given schemaQuery key
 * @param schemaQueryKey Key for the charts map based on a schemaQuery
 */
export function getCharts(schemaQueryKey: string) : List<DataViewInfo> {
    return getGlobalState('charts').get(schemaQueryKey);
}

/**
 * Sets the global state list of chart DataViewInfos for a given schemaQuery key
 * @param schemaQueryKey Key for the charts map based on a schemaQuery
 * @param dataViewInfos List of DataViewInfo objects defining the charts for the given key
 */
export function updateCharts(schemaQueryKey: string, dataViewInfos: List<DataViewInfo>) {
    setGlobal({
        QueryGrid_charts: getGlobalState('charts').set(schemaQueryKey, dataViewInfos)
    });
}

function getSelectedState(
    dataIds: List<string>,
    selected: List<string>,
    maxRows: number,
    totalRows: number
): GRID_CHECKBOX_OPTIONS {

    const selectedOnPage: number = dataIds.filter((id) => selected.indexOf(id) !== -1).size,
        totalSelected: number = selected.size;

    if (
        maxRows === selectedOnPage ||
        totalRows === totalSelected && totalRows !== 0 ||
        selectedOnPage === totalSelected && selectedOnPage === dataIds.size && selectedOnPage > 0
    ) {
        return GRID_CHECKBOX_OPTIONS.ALL;
    }
    else if (totalSelected > 0) {
        // if model has any selected show checkbox as indeterminate
        return GRID_CHECKBOX_OPTIONS.SOME;
    }

    return GRID_CHECKBOX_OPTIONS.NONE;
}

interface IGridSelectionResponse {
    selectedIds: List<any>
}

/**
 * Update model data with select changes
 * @param model
 * @param response
 */
export function updateSelections(model: QueryGridModel, response: IGridSelectionResponse)  {
    const selectedIds = response.selectedIds;
    const id = model.getId(),
        selectedLoaded: any = true;

    if (selectedIds !== undefined && selectedIds.size) {
        const { dataIds, maxRows, totalRows } = model;
        const selectedState = getSelectedState(dataIds, selectedIds, maxRows, totalRows);
        const updatedState = {
            selectedIds,
            selectedLoaded,
            selectedQuantity: selectedIds.size,
            selectedState
        } as any;

        setGlobal({
            QueryGrid_models: getGlobalState('models').set(model.getId(), model.merge(updatedState))
        });
    }
    else {
        setGlobal({
            QueryGrid_models: getGlobalState('models').set(id, model.merge({selectedLoaded}))
        });
    }
}

/**
 * Get the latest EditorModel object from the global state for a given modelId.
 * @param modelId QueryGridModel id to fetch
 */
export function getEditorModel(modelId: string): EditorModel {
    return getGlobalState('editors').get(modelId);
}

/**
 * Helper function for all callers/actions that would like to update information for an EditorModel in the global state.
 * @param model EditorModel in the global state to be updated, or to be added to global state if it does not already exist by Id
 * @param updates JS Object with the key/value pairs for updates to make to the model
 * @param failIfNotFound Boolean indicating if an error should be thrown if the model is not found in global state
 */
export function updateEditorModel(model: EditorModel, updates: any, failIfNotFound: boolean = true): EditorModel {
    if (failIfNotFound && !getGlobalState('editors').has(model.id)) {
        throw new Error('Unable to find EditorModel for modelId: ' + model.id);
    }

    const updatedModel = model.merge(updates) as EditorModel;

    setGlobal({
        QueryGrid_editors: getGlobalState('editors').set(model.id, updatedModel)
    });

    return updatedModel;
}

/**
 * Get the latest LookupStore object from the global state for a given QueryColumn.
 * @param col QueryColumn to fetch
 */
export function getLookupStore(col: QueryColumn): LookupStore {
    const key = LookupStore.key(col);
    return getGlobalState('lookups').get(key);
}

/**
 * Helper function for all callers/actions that would like to update information for a LookupStore in the global state.
 * @param store LookupStore in the global state to be updated, or to be added to global state if it does not already exist by col key
 * @param updates JS Object with the key/value pairs for updates to make to the store
 * @param failIfNotFound Boolean indicating if an error should be thrown if the store is not found in global state
 */
export function updateLookupStore(store: LookupStore, updates: any, failIfNotFound: boolean = true): LookupStore {
    if (failIfNotFound && !getGlobalState('lookups').has(store.key)) {
        throw new Error('Unable to find LookupStore for col: ' + store.key);
    }

    const updatedStore = store.merge(updates) as LookupStore;

    setGlobal({
        QueryGrid_lookups: getGlobalState('lookups').set(store.key, updatedStore)
    });

    return updatedStore;
}