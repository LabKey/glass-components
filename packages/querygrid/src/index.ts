/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { DataViewInfoTypes, EditorModel, getStateQueryGridModel, IDataViewInfo, SearchResultsModel } from './models';
import {
    addColumns,
    changeColumn,
    createQueryGridModelFilteredBySample,
    getFilterListFromQuery,
    getSelected,
    getSelectedData,
    getSelection,
    gridIdInvalidate,
    gridInit,
    gridInvalidate,
    gridRefresh,
    gridShowError,
    queryGridInvalidate,
    removeColumn,
    schemaGridInvalidate,
    searchUsingIndex,
    setSelected
} from './actions';
import {
    getEditorModel,
    getQueryGridModel,
    initQueryGridState,
    invalidateLineageResults,
    removeQueryGridModel,
    setQueryColumnRenderers,
    setQueryMetadata,
    invalidateProjectUsers
} from './global';
import {
    deleteRows,
    getQueryDetails,
    IImportData,
    importData,
    InsertFormats,
    InsertOptions,
    insertRows,
    InsertRowsResponse,
    invalidateQueryDetailsCacheKey,
    ISelectRowsResult,
    searchRows,
    selectRows,
    updateRows
} from './query/api';
import { MAX_EDITABLE_GRID_ROWS, NO_UPDATES_MESSAGE } from './constants';
import { buildQueryString, getLocation, Location, pushParameter, pushParameters, replaceParameters } from './util/URL';
import { URLResolver } from './util/URLResolver';
import { URLService } from './util/URLService';
import {
    AppRouteResolver,
    AssayResolver,
    AssayRunResolver,
    ListResolver,
    SampleSetResolver,
    SamplesResolver
} from './util/AppURLResolver';
import { QueryGridPanel } from './components/QueryGridPanel';
import { EditableGridPanel } from './components/editable/EditableGridPanel';
import { EditableGridPanelForUpdate } from './components/editable/EditableGridPanelForUpdate';
import { EditableGridLoader } from './components/editable/EditableGridLoader';
import { EditableGridLoaderFromSelection } from './components/editable/EditableGridLoaderFromSelection';
import { EditableGridModal } from './components/editable/EditableGridModal';
import { EditableColumnMetadata } from './components/editable/EditableGrid';
import { AliasRenderer } from './renderers/AliasRenderer';
import { AppendUnits } from './renderers/AppendUnits';
import { DefaultRenderer } from './renderers/DefaultRenderer';
import { FileColumnRenderer } from './renderers/FileColumnRenderer';
import { MultiValueRenderer } from './renderers/MultiValueRenderer';
import { BulkUpdateForm } from './components/forms/BulkUpdateForm';
import { QueryInfoForm } from './components/forms/QueryInfoForm';
import { LabelOverlay } from './components/forms/LabelOverlay';
import { LookupSelectInput } from './components/forms/input/LookupSelectInput';
import { SelectInput } from './components/forms/input/SelectInput';
import { QuerySelect } from './components/forms/QuerySelect';
import { PageDetailHeader } from './components/forms/PageDetailHeader';
import { DetailEditing } from './components/forms/detail/DetailEditing';
import { resolveDetailRenderer } from './components/forms/detail/DetailEditRenderer';
import { Detail } from './components/forms/detail/Detail';
import { handleInputTab, handleTabKeyOnTextArea, getProjectUsers } from './components/forms/actions';
import { IUser } from './components/forms/model';
import { FormStep, FormTabs, withFormSteps, WithFormStepsProps } from './components/forms/FormStep';
import { PlacementType } from './components/editable/Controls';
import { SchemaListing } from './components/listing/SchemaListing';
import { QueriesListing } from './components/listing/QueriesListing';
import { HeatMap } from './components/heatmap/HeatMap';
import  {
    addDateRangeFilter,
    last12Months,
    monthSort
} from './components/heatmap/utils';
import { SampleInsertPanel } from './components/samples/SampleInsertPanel';
import { SampleDeleteConfirmModal } from './components/samples/SampleDeleteConfirmModal';
import { SearchResultCard } from './components/search/SearchResultCard';
import { SearchResultsPanel } from './components/search/SearchResultsPanel';
import {
    deleteSampleSet,
    getSampleDeleteConfirmationData,
    getSampleSet,
    loadSelectedSamples
} from './components/samples/actions';
import { SampleSetDeleteConfirmModal } from './components/samples/SampleSetDeleteConfirmModal';
import { SampleSetDetailsPanel } from './components/samples/SampleSetDetailsPanel';
import { AssayImportPanels } from './components/assay/AssayImportPanels';
import { BatchPropertiesPanel } from './components/assay/BatchPropertiesPanel';
import { RunPropertiesPanel } from './components/assay/RunPropertiesPanel';
import { RunDataPanel } from './components/assay/RunDataPanel';
import { AssayUploadGridLoader } from './components/assay/AssayUploadGridLoader';
import { AssayDesignDeleteConfirmModal } from './components/assay/AssayDesignDeleteConfirmModal';
import { AssayResultDeleteConfirmModal } from './components/assay/AssayResultDeleteConfirmModal';
import { AssayRunDeleteConfirmModal } from './components/assay/AssayRunDeleteConfirmModal';
import { AssayImportSubMenuItem } from './components/assay/AssayImportSubMenuItem';
import {
    AssayUploadResultModel,
    AssayWizardModel,
    IAssayUploadOptions,
    IAssayURLContext
} from './components/assay/models';
import {
    deleteAssayDesign,
    deleteAssayRuns,
    getBatchPropertiesModel,
    getBatchPropertiesRow,
    getImportItemsForAssayDefinitions,
    getRunPropertiesModel,
    getRunPropertiesRow,
    importAssayRun,
    uploadAssayRunFiles
} from './components/assay/actions';
import { PreviewGrid } from './components/PreviewGrid';
import { flattenBrowseDataTreeResponse, ReportURLMapper, } from './components/report-list/model';
import { ReportItemModal, ReportList, ReportListItem, ReportListProps } from './components/report-list/ReportList';
import { LINEAGE_GROUPING_GENERATIONS } from './components/lineage/constants'
import { LineageFilter } from './components/lineage/models'
import { VisGraphNode } from './components/lineage/vis/VisGraphGenerator'
import { LineageGraph } from './components/lineage/LineageGraph';
import { LineageGrid } from './components/lineage/LineageGrid';
import { SampleTypeLineageCounts } from './components/lineage/SampleTypeLineageCounts';
import { OmniBox } from './components/omnibox/OmniBox';
import { HeaderWrapper } from './components/navigation/HeaderWrapper';
import { NavigationBar } from './components/navigation/NavigationBar';
import { NavItem } from './components/navigation/NavItem';
import { MenuSectionConfig } from './components/navigation/ProductMenuSection';
import { ITab, SubNav } from './components/navigation/SubNav';
import { Breadcrumb } from './components/navigation/Breadcrumb';
import { BreadcrumbCreate } from './components/navigation/BreadcrumbCreate';
import { MenuSectionModel, MenuItemModel, ProductMenuModel } from './components/navigation/model';
import { UserSelectInput } from './components/forms/input/UserSelectInput';

export {
    // global state functions
    initQueryGridState,
    getStateQueryGridModel,
    getQueryGridModel,
    getEditorModel,
    removeQueryGridModel,
    setQueryMetadata,
    setQueryColumnRenderers,

    // grid functions
    getSelected,
    getSelectedData,
    getSelection,
    getFilterListFromQuery,
    gridInit,
    gridInvalidate,
    gridIdInvalidate,
    queryGridInvalidate,
    schemaGridInvalidate,
    gridRefresh,
    gridShowError,

    // query related items
    ISelectRowsResult,
    InsertRowsResponse,
    IImportData,
    InsertFormats,
    InsertOptions,
    insertRows,
    selectRows,
    searchRows,
    updateRows,
    deleteRows,
    importData,
    getQueryDetails,
    invalidateQueryDetailsCacheKey,
    getSampleDeleteConfirmationData,
    setSelected,

    // editable grid related items
    addColumns,
    changeColumn,
    removeColumn,
    MAX_EDITABLE_GRID_ROWS,
    NO_UPDATES_MESSAGE,
    EditableGridLoaderFromSelection,
    EditableGridLoader,

    // location related items
    Location,
    URLResolver,
    URLService,
    AppRouteResolver,
    AssayResolver,
    AssayRunResolver,
    ListResolver,
    SamplesResolver,
    SampleSetResolver,
    getLocation,
    pushParameter,
    pushParameters,
    replaceParameters,
    buildQueryString,

    // renderers
    AliasRenderer,
    AppendUnits,
    DefaultRenderer,
    FileColumnRenderer,
    MultiValueRenderer,
    resolveDetailRenderer,

    // components
    LabelOverlay,
    EditableGridPanel,
    EditableGridPanelForUpdate,
    EditableGridModal,
    QueryGridPanel,
    PreviewGrid,
    BulkUpdateForm,
    QueryInfoForm,
    LookupSelectInput,
    SelectInput,
    QuerySelect,
    UserSelectInput,
    PageDetailHeader,
    DetailEditing,
    Detail,
    SchemaListing,
    QueriesListing,
    HeatMap,

    // interfaces
    EditableColumnMetadata,

    // types
    PlacementType,
    EditorModel,

    // user-related
    getProjectUsers,
    invalidateProjectUsers,
    IUser,

    // samples-related
    SampleInsertPanel,
    SampleDeleteConfirmModal,
    SampleSetDetailsPanel,
    SampleSetDeleteConfirmModal,
    deleteSampleSet,
    getSampleSet,
    createQueryGridModelFilteredBySample,
    loadSelectedSamples,

    // search-related
    SearchResultsModel,
    SearchResultCard,
    SearchResultsPanel,
    searchUsingIndex,

    // assay
    AssayUploadResultModel,
    AssayDesignDeleteConfirmModal,
    AssayResultDeleteConfirmModal,
    AssayRunDeleteConfirmModal,
    AssayWizardModel,
    IAssayURLContext,
    IAssayUploadOptions,
    AssayUploadGridLoader,
    AssayImportPanels,
    BatchPropertiesPanel,
    RunPropertiesPanel,
    RunDataPanel,
    AssayImportSubMenuItem,
    importAssayRun,
    uploadAssayRunFiles,
    deleteAssayDesign,
    deleteAssayRuns,
    getImportItemsForAssayDefinitions,
    getRunPropertiesModel,
    getRunPropertiesRow,
    getBatchPropertiesModel,
    getBatchPropertiesRow,

    // forms
    handleInputTab,
    handleTabKeyOnTextArea,
    withFormSteps,
    WithFormStepsProps,
    FormStep,
    FormTabs,

    // heatmap
    addDateRangeFilter,
    last12Months,
    monthSort,

    // DataViewInfo
    DataViewInfoTypes,
    IDataViewInfo,

    // report-list
    flattenBrowseDataTreeResponse,
    ReportURLMapper,
    ReportListItem,
    ReportItemModal,
    ReportListProps,
    ReportList,

    // lineage
    LINEAGE_GROUPING_GENERATIONS,
    LineageFilter,
    LineageGraph,
    LineageGrid,
    SampleTypeLineageCounts,
    VisGraphNode,
    invalidateLineageResults,

    // OmniBox components
    OmniBox,

    // Navigation types
    MenuSectionConfig,
    ProductMenuModel,
    MenuSectionModel,
    MenuItemModel,

    // Navigation components
    HeaderWrapper,
    ITab,
    NavItem,
    NavigationBar,
    SubNav,
    Breadcrumb,
    BreadcrumbCreate
}