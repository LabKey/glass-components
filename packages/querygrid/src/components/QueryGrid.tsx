/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'reactn'
import { List, Map } from 'immutable'
import {
    GRID_CHECKBOX_OPTIONS, GRID_SELECTION_INDEX, QueryGridModel, SchemaQuery,
    Grid, GridColumn, GridProps, Alert, LoadingSpinner, generateId
} from '@glass/base'

import { QUERY_GRID_PREFIX } from '../constants'
import { gridInit, toggleGridRowSelection, toggleGridSelected, sort, reloadQueryGridModel } from '../actions'
import { getStateQueryGridModel, getStateModelId } from '../model'
import { headerCell, headerSelectionCell } from '../renderers'
import { getBrowserHistory } from "../util/global";

interface QueryGridProps {
    model?: QueryGridModel
    schemaQuery?: SchemaQuery
}

interface QueryGridState {
    /**
     * modelId can be used to maintain a single instance model. This is useful when an explicit model
     * is not provided but persistence of the generated model is desired (fetched results, etc)
     */
    modelId: string

    /**
     * Function returned by the getBrowserHistory().listen() call so that we can cleanup after unmount
     */
    unlisten: any
}

export class QueryGrid extends React.Component<QueryGridProps, QueryGridState> {

    constructor(props: QueryGridProps) {
        super(props);

        // bind event handlers
        this.headerCell = this.headerCell.bind(this);
        this.selectAll = this.selectAll.bind(this);
        this.sort = this.sort.bind(this);

        const { model, schemaQuery } = this.props;
        let _modelId;

        if (model) {
            _modelId = model.getId();
        }
        else {
            if (!schemaQuery) {
                throw new Error('QueryGrid: If a model is not provided, a SchemaQuery is required.');
            }

            _modelId = generateId(QUERY_GRID_PREFIX);
        }

        const unlisten = getBrowserHistory().listen((location, action) => {
            if (this.props.model && this.props.model.bindURL) {
                reloadQueryGridModel(this.props.model);
            }
        });

        // set local state for this component
        this.state = {
            modelId: _modelId,
            unlisten
        };
    }

    componentDidMount() {
        this.initModel(this.props);
    }

    componentWillReceiveProps(nextProps: QueryGridProps) {
        this.initModel(nextProps);
    }

    componentWillUnmount() {
        const { unlisten } = this.state;
        if (unlisten) {
            unlisten();
        }
    }

    initModel(props: QueryGridProps) {
        const { model, schemaQuery } = props;
        const { modelId } = this.state;

        if (model && !model.isLoaded && !model.isLoading) {
            gridInit(model);
        }
        else if (!this.getModel(props)) {
            gridInit(getStateQueryGridModel(modelId, schemaQuery));
        }
    }

    headerCell(column: GridColumn, i: number) {
        const model = this.getModel(this.props);

        if (model.allowSelection && column.index.toLowerCase() === GRID_SELECTION_INDEX && !model.editable) {
            return headerSelectionCell(this.selectAll, model.selectedState, !(model.isLoaded && model.totalRows !== 0));
        }

        return headerCell(this.sort, column, i, model.allowSelection, model.sortable);
    }

    getColumns(): List<any> {
        const model = this.getModel(this.props);

        if (model.editable) {
            return model.getInsertColumns();
        }

        if (model.allowSelection) {
            const selColumn = new GridColumn({
                index: GRID_SELECTION_INDEX,
                title: '&nbsp;',
                showHeader: true,
                cell: (selected: boolean, row) => {
                    return <input
                        checked={selected === true}
                        type="checkbox"
                        onChange={this.select.bind(this, row)}/>;
                }
            });

            return List([selColumn]).concat(model.getDisplayColumns()).toList();
        }

        return model.getDisplayColumns();
    }

    getModel(props: QueryGridProps): QueryGridModel {
        const { model, schemaQuery } = props;
        const { modelId } = this.state;

        // if a model is explicitly defined, use the id from it
        const stateModelId = model ? model.getId() : getStateModelId(modelId, schemaQuery);

        // need to access this.global directly to connect this component to the re-render cycle
        return this.global.QueryGrid_models.get(stateModelId);
    }

    selectAll(evt) {
        const model = this.getModel(this.props);

        if (model) {
            const selected = evt.currentTarget.checked === true && model.selectedState !== GRID_CHECKBOX_OPTIONS.SOME;
            toggleGridSelected(model, selected);
        }
    }

    select(row: Map<string, any>, evt) {
        const model = this.getModel(this.props);

        if (model) {
            toggleGridRowSelection(model, row, evt.currentTarget.checked === true);
        }
    }

    sort(gridColumn, dir) {
        const model = this.getModel(this.props);

        if (model) {
            sort(model, gridColumn.index, dir);
        }
    }

    render() {
        const model = this.getModel(this.props);
        if (!model) {
            return (
                <div>
                    <LoadingSpinner/>
                </div>
            )
        }
        else if (model.isError) {
            return <Alert>{model.message ? model.message : 'Something went wrong loading the QueryGridModel.'}</Alert>
        }

        const gridProps: GridProps = {
            calcWidths: true,
            columns: this.getColumns(),
            condensed: true,
            data: model.getData(),
            gridId: model.getId(),
            headerCell: this.headerCell,
            isLoading: model.isLoading,
            loadingText: <LoadingSpinner/>
        };

        return <Grid {...gridProps} />
    }
}