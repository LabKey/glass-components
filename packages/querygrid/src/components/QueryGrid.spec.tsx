import React from 'reactn';
import renderer from 'react-test-renderer';

import { QueryGrid } from "./QueryGrid";
import { QueryGridModel, SchemaQuery } from "@glass/models/src";
import { initQueryGridState, updateQueryGridModel } from "../global";

beforeAll(() => {
    initQueryGridState()
});

// Mock all the actions to test just the rendering parts for QueryGrid itself
jest.mock('../actions');

describe("QueryGrid render", () => {
    test("loading", () => {

        const schemaQuery = new SchemaQuery({
            schemaName: "schema",
            queryName: "q-snapshot"
        });
       const tree = renderer.create(<QueryGrid schemaQuery={schemaQuery}/>).toJSON();
       expect(tree).toMatchSnapshot();
    });

    test("query grid error", () => {
        const modelId = "queryGridError";
        const schemaQuery = new SchemaQuery({
            schemaName: "schema",
            queryName: "q-snapshot"
        });
        const model = new QueryGridModel({
            id: modelId,
            isLoaded: true,
            isLoading: false,
            isError: true,
            schema: schemaQuery.schemaName,
            query: schemaQuery.queryName,
        });
        updateQueryGridModel(model, {}, false);
        const tree = renderer.create(<QueryGrid model={model} schemaQuery={schemaQuery}/>);
        expect(tree).toMatchSnapshot();
    });

    test("allow selection", () => {
        const modelId = "gridWithSelection";
        const schemaQuery = new SchemaQuery({
            schemaName: "schema",
            queryName: "withSelection"
        });
        const model = new QueryGridModel({
            id: modelId,
            isLoaded: true,
            schema: schemaQuery.schemaName,
            query: schemaQuery.queryName,

        });
        // TODO make sure selection column shows up
        // TODO ??? click and verify state
    });

});