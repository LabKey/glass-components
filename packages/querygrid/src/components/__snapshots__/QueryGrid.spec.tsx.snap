// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render loading 1`] = `
<div>
  <span
    className=""
  >
    <i
      aria-hidden="true"
      className="fa fa-spinner fa-spin"
    />
     
    Loading...
  </span>
</div>
`;

exports[`render query grid error 1`] = `
<div
  className="alert alert-danger"
  role="alert"
>
  Something went wrong loading the QueryGridModel.
</div>
`;
